fish-rust/src/fish.rs

```diff
@@ -0,0 +1,877 @@
+use std::ffi::{CString, OsStr, OsString};
+use std::fs::File;
+use std::io;
+use std::mem::MaybeUninit;
+use std::os::unix::prelude::*;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+// FIXME: when the crate is actually called fish and not fish-rust, read this from cargo
+// See: https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates
+// for reference
+const PACKAGE_NAME: &str = "fish"; // env!("CARGO_PKG_NAME");
+const DOC_DIR: &str = {
+    match option_env!("DOCDIR") {
+        Some(e) => e,
+        None => "share/doc/fish",
+    }
+};
+const DATA_DIR: &str = {
+    match option_env!("DATADIR") {
+        Some(e) => e,
+        None => "share",
+    }
+};
+const SYSCONF_DIR: &str = {
+    match option_env!("SYSCONFDIR") {
+        Some(e) => e,
+        None => "/etc",
+    }
+};
+const BIN_DIR: &str = {
+    match option_env!("BINDIR") {
+        Some(e) => e,
+        None => "bin",
+    }
+};
+
+// C++ had this as optional, and used CMAKE_BINARY_DIR,
+// should probably be swapped to `OUT_DIR` once CMake is gone?
+// const OUT_DIR: &str = env!("OUT_DIR", "OUT_DIR was not specified");
+const OUT_DIR: &str = env!("FISH_BUILD_DIR");
+
+/// container to hold the options specified within the command line
+#[derive(Default, Debug)]
+struct FishCmdOpts {
+    /// Future feature flags values string
+    features: WString,
+    /// File path for debug output.
+    debug_output: Option<OsString>,
+    /// File path for profiling output, or empty for none.
+    profile_output: Option<OsString>,
+    profile_startup_output: Option<OsString>,
+    /// Commands to be executed in place of interactive shell.
+    batch_cmds: Vec<OsString>,
+    /// Commands to execute after the shell's config has been read.
+    postconfig_cmds: Vec<OsString>,
+    /// Whether to print rusage-self stats after execution.
+    print_rusage_self: bool,
+    /// Whether no-config is set.
+    no_config: bool,
+    /// Whether no-exec is set.
+    no_exec: bool,
+    /// Whether this is a login shell.
+    is_login: bool,
+    /// Whether this is an interactive session.
+    is_interactive_session: bool,
+    /// Whether to enable private mode.
+    enable_private_mode: bool,
+}
+
+/// \return a timeval converted to milliseconds.
+#[allow(clippy::unnecessary_cast)]
+fn tv_to_msec(tv: &libc::timeval) -> i64 {
+    // milliseconds per second
+    let mut msec = tv.tv_sec as i64 * 1000;
+    // microseconds per millisecond
+    msec += tv.tv_usec as i64 / 1000;
+    msec
+}
+
+fn print_rusage_self(mut dest: impl io::Write) {
+    let mut rs = MaybeUninit::zeroed();
+    if unsafe { libc::getrusage(libc::RUSAGE_SELF, rs.as_mut_ptr()) } != 0 {
+        let s = CString::new("getrusage").unwrap();
+        unsafe { libc::perror(s.as_ptr()) }
+        return;
+    }
+    let rs: libc::rusage = unsafe { rs.assume_init() };
+    let rss_kb = if cfg!(target_vendor = "apple") {
```

Comment:
I'm pretty sure that we can use `target_os = "macos"` which is more consistent with the other Rust

Reply:
yeah if we ever need that we can revisit all of the target_os directives

Reply:
This would actually have introduced a bug for our (potential) iOS users, since it is kB on iOS https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/getrusage.2.html